proposals/NNNN-hashable-enhancements.md

```diff
@@ -0,0 +1,950 @@
+}
+```
+
+[SE-0185] did not specify a hash function to be used for such
+conformances, leaving it as an implementation detail of the compiler
+and the standard library. Doing this well requires the use of a hash
+function that works equally well on any number of components,
+regardless of their expected distributions.
+
+[SE-0185]: https://github.com/apple/swift-evolution/blob/master/proposals/0185-synthesize-equatable-hashable.md
+
+Luckily, this problem has occured in other contexts before, and there
+is an extensive list of hash functions that have been designed for
+exactly such cases: [Foller-Noll-Vo][FNV-1a], [MurmurHash],
+[CityHash], [SipHash], and [HighwayHash] are just a small selection of
+these. (The last two algorithms have some light cryptographic elements
+so that they provide reasonable level of protection against hash
+collision attacks.  This makes them better-suited for use in
+general-purpose hashed collections like `Set` and `Dictionary`.)
+
+[FNV-1a]: http://www.isthe.com/chongo/tech/comp/fnv/index.html
+[MurmurHash]: https://github.com/aappleby/smhasher
+[CityHash]: https://github.com/google/cityhash
+[SipHash]: https://131002.net/siphash/
+[HighwayHash]: https://github.com/google/highwayhash
+
+Since [SE-0185] required the standard library to implement a
+high-quality universal hash function, it seems like a good idea to
+expose it as public API, so that manual `Hashable` implementations can
+take advantage of it, too.
+
+Universal hash functions work by maintaining some internal state --
+this can be as simple as a single 32/64-bit integer value (for
+e.g. [FNV-1a]), but it is usually much wider than that. For example,
+[SipHash] maintains a state of 256 bits, while [HighwayHash] uses 1024
+bits. 
+
+## <a name="proposed-solution">Proposed solution</a>
+
+We solve `Hashable`'s implementation problems in two parts. First, we
+make the standard library's hash function public. Second, we replace
+`hashValue` with a requirement that is designed specifically to
+eliminate the guesswork from manual `Hashable` implementations.
+
+### <a name="hasher">The `Hasher` struct</a>
+
+We propose to expose the standard library's standard hash function as
+a new, public struct type, called `Hasher`. This new struct captures
+the state of the hash function, and provides the following operations:
+
+1. An initializer to create an empty state. To make hash values less
+   predictable, the standard hash function uses an arbitrary 128-bit
+   seed value, specified as a tuple of two `UInt64` integers.
+   `Hasher` uses a per-execution random seed value by default;
+   however, we also allow programmers to specify their own seed
+   values, primarily to enable per-instance random seeding in their
+   custom hashing collections.
+   
+    ```swift
+    public struct Hasher {
+      public init()  // Use the default per-execution random seed value
+      public init(seed: (UInt64, UInt64)) // Combines `seed` with the default seed
```

Comment:
If it combines with the default seed (which is per-execution random), then the caller cannot choose a seed, only modify a random one, or am I misunderstanding this?

Otherwise looks good. As further enhancement, I'd love to be able to ignore certain members for inclusion in the compiler-generated `Hashable` support...

Reply:
Great point! I changed the meaning of the seed in a revision of this document, but forgot to make it consistent with the text. I think it's best not to let Swift programs disable the random seed, because that may lead to subtle dependencies on `Hasher`'s algorithm. I'm clarifying this in a new commit.

Reply:
One of the reasonable situations where one might want to have hashes be deterministically and persistently seeded over executions would be this:

Say you're implementing a [Bloom Filter](https://en.wikipedia.org/wiki/Bloom_filter) or some other data structure making use of hashes and want to write unit tests testing specific behavior.

One of the features of a bloom filter is that while they cannot tell you which or how many elements they contain (only which they don't), one can calculate an approximation of the number of elements. In a unit tests one might thus want to fill the bloom filter with a pre-defined list of elements and check whether the approximated element count matches the expected value:

```swift
let hashers = (Hasher(seed: 13), Hasher(seed: 37))
var bloomFilter = BloomFilter(capacity: 42, falsePositiveRate: 0.1, hashers: hashers)
for element in ["lorem", "ipsum", "dolor", "sit", "amet"] {
    bloomFilter.insert(element)
}
XCAssertEqual(bloomFilter.approximateCount, <some value>, <some accurracy>)
```

With randomization in place this value can fluctuate greatly and thus turn your tests into some kind of heisentests.